storage/src/mmr/hasher.rs

```diff
@@ -1,82 +1,212 @@
+        peak_digests: impl Iterator<Item = &'b H::Digest>,
     ) -> H::Digest {
         self.update_with_pos(pos);
-        for hash in peak_hashes {
-            self.update_with_hash(hash);
+        for digest in peak_digests {
+            self.update_with_digest(digest);
         }
-        self.finalize_reset()
+        self.finalize()
     }
 
-    pub(crate) fn update_with_pos(&mut self, pos: u64) {
-        self.hasher.update(&pos.to_be_bytes());
+    fn digest(&mut self, data: &[u8]) -> H::Digest {
+        self.hasher.update(data);
+        self.finalize()
     }
-    pub(crate) fn update_with_hash(&mut self, hash: &H::Digest) {
-        self.hasher.update(hash.as_ref());
+}
+
+/// Hasher for computing leaf, node and root digests when the tree is being grafted onto another
+/// MMR.
+pub struct Grafting<'a, H: CHasher, S: Storage<H::Digest>> {
+    hasher: Standard<'a, H>,
+    height: u32,
+    base_mmr: &'a S,
+}
+
+impl<'a, H: CHasher, S: Storage<H::Digest>> Grafting<'a, H, S> {
+    pub fn new(hasher: &'a mut H, height: u32, base_mmr: &'a S) -> Self {
+        Self {
+            hasher: Standard::new(hasher),
+            height,
+            base_mmr,
+        }
     }
-    pub(crate) fn update_with_element(&mut self, element: &[u8]) {
-        self.hasher.update(element);
+
+    /// Access the underlying [Standard] (non-grafting) hasher.
+    pub fn standard(&mut self) -> &mut Standard<'a, H> {
+        &mut self.hasher
     }
-    pub(crate) fn finalize_reset(&mut self) -> H::Digest {
-        self.hasher.finalize()
+
+    /// Compute the position of the leaf in the base tree onto which we should graft the leaf at
+    /// position `pos` in the source tree.
+    ///
+    /// This position is computed by walking up the MMR exactly self.height steps from the leaf.
+    /// Since we don't know exactly where in the subtree this leaf falls, we map it to the
+    /// right-most leaf in the subtree and walk up from there. Runtime is O(log2(n)) in the given
+    /// position.
+    fn destination_pos(&self, pos: u64) -> u64 {
+        let peak_leaf_num = leaf_pos_to_num(pos).unwrap();
+        let chunk_size_bits = 1 << self.height;
+
+        // The rightmost-leaf in the corresponding segment of the peak tree:
+        let base_leaf_num = (peak_leaf_num * chunk_size_bits) + chunk_size_bits - 1;
+
+        // Walking up self.height levels from the rightmost leaf involves simply adding self.height
+        // to its position.
+        leaf_num_to_pos(base_leaf_num) + self.height as u64
+    }
+}
+
+impl<H: CHasher, S: Storage<H::Digest>> Hasher<H> for Grafting<'_, H, S> {
+    async fn leaf_digest(&mut self, pos: u64, element: &[u8]) -> Result<H::Digest, Error> {
+        let base_node_pos = self.destination_pos(pos);
+        if base_node_pos >= self.base_mmr.size() {
+            // The base tree doesn't yet have a node where we can graft this position.
+            return self.hasher.leaf_digest(pos, element).await;
```

Comment:
This will make for some interesting proof logic 👀

Reply:
It's not too bad: you just need to include the digest of any peak whose height is less than the grafting height.

Reply:
(This will be documented in the next PR)